onnxruntime/core/providers/xnnpack/nn/unpool.cc

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "unpool.h"
+#include <algorithm>
+
+#include "core/common/common.h"
+#include "core/common/inlined_containers_fwd.h"
+#include "core/framework/tensor_shape.h"
+#include "core/graph/graph.h"
+#include "core/providers/cpu/nn/pool_attributes.h"
+#include "core/providers/utils.h"
+#include "core/framework/tensorprotoutils.h"
+
+// to sanity check output shape
+#include "core/framework/tensorprotoutils.h"
+#include "gsl/gsl-lite.hpp"
+
+namespace onnxruntime {
+namespace xnnpack {
+namespace {
+
+TensorShapeVector InferOutputSizeForUnPool(const PoolAttributes& pool_attrs,
+                                           const TensorShape& input_shape) {
+  // Calculate output tensor shape from attributes
+  TensorShapeVector inferred_output_dims(input_shape.NumDimensions());
+
+  // Copy batch and channel dims
+  inferred_output_dims[0] = input_shape[0];
+  inferred_output_dims[3] = input_shape[3];
+
+  // For feature dims calculate reversing the formula used for MaxPool
+  for (size_t dim = 0; dim < pool_attrs.kernel_shape.size(); ++dim) {
+    inferred_output_dims[dim + 1] =
+        (input_shape[dim + 1] - 1) * pool_attrs.strides[dim] -
+        (pool_attrs.pads[dim] + pool_attrs.pads[pool_attrs.kernel_shape.size() + dim]) +
+        pool_attrs.kernel_shape[dim];
+  }
+
+  return inferred_output_dims;
+}
+}  // namespace
+
+// MaxUnpool doesn't have any quantization params
+bool MaxUnpool::IsOnnxNodeSupported(const NodeUnit& node_unit,
+                                    const GraphViewer& graph_viewer) {
+  ORT_UNUSED_PARAMETER(graph_viewer);
```

Comment:
nit: comment out or omit the `graph_viewer` in the args - e.g. `const GraphViewer& /*graph_viewer*/` and you won't need the ORT_UNUSED_PARAMETER.

Reply:
Removed and add more comments on the thrid inputs to clarify the hard requirement.